akka-remote/src/main/scala/akka/remote/artery/ArteryTcpTransport.scala

```diff
@@ -0,0 +1,352 @@
+  import ArteryTransport._
+  import FlightRecorderEvents._
+
+  @volatile private var inboundKillSwitch: SharedKillSwitch = KillSwitches.shared("inboundKillSwitch")
+  private var serverBinding: Option[Future[ServerBinding]] = None
+
+  private val sslEngineProvider: OptionVal[SSLEngineProvider] =
+    // FIXME load from config
+    if (tlsEnabled) OptionVal.Some(new ConfigSSLEngineProvider(system))
+    else OptionVal.None
+
+  override protected def startTransport(): Unit = {
+    // FIXME new TCP events
+    topLevelFREvents.loFreq(Transport_AeronStarted, NoMetaData)
+  }
+
+  override protected def outboundTransportSink(
+    outboundContext: OutboundContext,
+    streamId:        Int,
+    bufferPool:      EnvelopeBufferPool): Sink[EnvelopeBuffer, Future[Done]] = {
+    implicit val sys = system
+
+    val host = outboundContext.remoteAddress.host.get
+    val port = outboundContext.remoteAddress.port.get
+    val remoteAddress = InetSocketAddress.createUnresolved(host, port)
+    val connectionTimeout = 5.seconds // FIXME config
+
+    def connectionFlow: Flow[ByteString, ByteString, Future[Tcp.OutgoingConnection]] =
+      if (tlsEnabled) {
+        Tcp().outgoingTlsConnectionWithSSLEngine(
+          remoteAddress,
+          createSSLEngine = () ⇒ sslEngineProvider.get.createClientSSLEngine(),
+          connectTimeout = connectionTimeout)
+      } else {
+        Tcp()
+          .outgoingConnection(
+            remoteAddress,
+            halfClose = true, // issue https://github.com/akka/akka/issues/24392 if set to false
+            connectTimeout = connectionTimeout)
+      }
+
+    def connectionFlowWithRestart: Flow[ByteString, ByteString, NotUsed] = {
+      val flowFactory = () ⇒ connectionFlow.mapMaterializedValue(_ ⇒ NotUsed)
+        .log(name = s"outbound connection to [${outboundContext.remoteAddress}], ${streamName(streamId)} stream")
+        .addAttributes(Attributes.logLevels(onElement = LogLevels.Off, onFailure = Logging.WarningLevel))
+
+      // FIXME config of backoff
+      import scala.concurrent.duration._
+      if (streamId == ControlStreamId) {
+        // restart of inner connection part important in control flow, since system messages
+        // are buffered and resent from the outer SystemMessageDelivery stage.
+        // FIXME The mat value Future is currently never completed, because RestartFlow will retry forever, should it give up?
+        //       related to give-up-system-message-after ?
+        RestartFlow.withBackoff[ByteString, ByteString](1.second, 5.seconds, 0.1)(flowFactory)
+      } else {
+        // Best effort retry a few times
+        // FIXME only restart on failures?, but missing in RestartFlow, see https://github.com/akka/akka/pull/23911
+        RestartFlow.withBackoff[ByteString, ByteString](1.second, 5.seconds, 0.1, maxRestarts = 3)(flowFactory)
+      }
+    }
+
+    Flow[EnvelopeBuffer]
+      .map { env ⇒
+        val bytes = ByteString(env.byteBuffer)
```

Comment:
seems a shame that every msg will involve an array copy but i guess that can't be avoided given we re-use the buffer

Reply:
I agree. In the end we have to copy it to a `Array[Byte]` for deserialization unless the serializer supports ByteBuffer (which is a new thing for Artery). Would be nice if we could at least copy it to a ByteString backed by a single `Array[Byte]` instead of the `Vector` based ByteString. Then no further copies when deserializing. (not completely safe since arrays are mutable, but as internal api/usage we could have such ByteString.

Adding a TODO comment here.

Reply:
sorry, this was for outbound and my thoughts were mixed up with inbound